Create the Text component and style guide variables

jbirori opened the pull request:
Created the Text component, defined spacing + color variables, and did some general clean up.

Usage of the text component is as follows:

```
import { Text } from "../components"
...
<Text text="This is the primary Heading" type="H1" />
<Text text="This is the large Body text with red color" style={{color: 'red'}} type="p+" />
```

<img width="828" alt="Screen Shot 2022-12-29 at 1 19 20 AM" src="https://user-images.githubusercontent.com/25678578/209889003-2739c3f4-b461-4321-9aac-3f2611ed603f.png">

jbirori committed:
Create the Text component, define spacing and color variables, and general clean up

MagicWilliams commented on src/styles/styles.scss:
```diff
@@ -23,7 +32,7 @@ body {
 .App {
   font-family: "Omnes", sans-serif;
-  color: #fff;
+  color: $color_white;
```
can we rename these variables to just be `$black` `$white`?

MagicWilliams commented on src/components/Text/types.ts:
```diff
@@ -0,0 +1,26 @@
+import React, { ReactNode } from "react";
+
+type HeadingTextTypes = 'H1' | 'H2' | 'H3' | 'H4' | 'H5' | 'H6';
+type BodyTextTypes = 'p+' | 'p' | 'p-';
+
+export type TextProps = {
+  /**
+   * Custom class(es) to apply
+   */
+  className?: string;
+
+  /**
+   * Custom styles to apply
+   */
+  style?: React.CSSProperties;
+
+  /**
+   * Main text content to render
+   */
+  text: ReactNode;
+
+  /**
+   * Predefined text type
+   */
+  type: HeadingTextTypes | BodyTextTypes;
+};
```
can we put this file in a `types` subfolder?

jbirori replied:
I find it easier to navigate supporting files for components when they are all located in the same place. E.g. im working on X component and I know i can find all the styles/types/util functions etc within the X directory. As we add more types i agree a higher level types folder would be useful for ones that are used more globally!

MagicWilliams replied:
fashooo - could we at least name this file something other than types.ts (perhaps ComponentName.ts) so we can avoid having lots of files with the same name?

MagicWilliams commented on src/pages/index.tsx:
```diff
@@ -1,8 +1,8 @@
+import { SEO } from "../components";
 import logo from "../images/dialup-retro-logo.png";
-import "../styles.scss";
+import "../styles/styles.scss";
```
what do you think about calling this global stylesheet something like `global.scss` instead of `styles/styles.scss`?

jbirori replied:
Here for it

MagicWilliams commented on src/components/Text/styles.scss:
```diff
@@ -0,0 +1,74 @@
+  margin-bottom: $spacing_xx_small;
+}
+
+.H1 {
```
nit but could we lowercase these class selectors?

MagicWilliams replied:
also what do you think about selecting the elements, i.e. via `.App h1 { styles }` in the global stylesheet as opposed to needing to designate a className for every element? I think that could help us when it comes to usage by just needing to use the right tag instead of passing text, classname, style, and type to a text component via props

jbirori replied:
Yeahh down to lowercase, was just following the style guide in Figma so that it matches

MagicWilliams commented on src/components/Text/styles.scss:
```diff
@@ -0,0 +1,74 @@
+@mixin BodyText {
+  color: $color_white;
+  font-family: "Operator Mono";
+  font-weight: 350;
```
I feel like I typically see font-weights in 100, 200, 300, 400, etc, is 350 a valid weight? j curious idk if i've seen it before?

jbirori replied:
Figma was showing the medium font weight as 350, could be a weird outlier 🤷🏽‍♂️

MagicWilliams commented on src/components/Text/styles.scss:
```diff
@@ -0,0 +1,74 @@
+  font-family: "Operator Mono";
+  font-weight: 350;
+  line-height: 125%;
+  margin-top: $spacing_xx_small;
```
could we name spacing variables like `$spacing-xs`, `$spacing-sm`, `$spacing-md`, `$spacing-lg`, etc?

jbirori replied:
Sure

MagicWilliams commented on src/components/Text/Text.tsx:
```diff
@@ -0,0 +1,25 @@
+import React from "react";
+import { TextProps } from "./types";
+import "./styles.scss";
```
since we have another styles.scss file, what do you think about naming the associated stylesheet after the Component it is for?

jbirori replied:
Down

jbirori replied:
Actually, i dont think it matters much. Was hoping we could maybe set a precedent where each component folder follows the same structure:

ComponentName [folder]
- ComponentName.tsx
- styles.scss
- types.ts
- utils.ts


What do you think?

MagicWilliams replied:
I just think when it comes to working on stuff having X number of styles and types folders will be inconvenient - naming the associated Types and Styles files after their component name would only require us to know the associated component and the extension of the file we're looking for which I think will be easier to maintain going fwd. Not that this site will even get too large but in terms of establishing a precedent I think it could be helpful

MagicWilliams replied:
I also think utils should live in the project root - I find when working on sites that oftentimes utility functions are used in several places across the app, so this would avoid us both A) needing to copy functions to different utils files, and B) needing to grab another component's utils file when needing the same function

jbirori replied:
For sure, so something like this?

ComponentName [folder]
* ComponentName.tsx
* ComponentNameStyles.scss
* ComponentNameTypes.ts
* ComponentNameUtils.ts

jbirori replied:
Yeah i definitely agree on the utils point, for global utils it makes sense for those to live in a global util file rather than copy pasting to multiple files. Although to avoid super long files as the project grows, I think some degree of isolation can be helpful when things are only used for specific components

MagicWilliams commented on src/styles/styles.scss:
```diff
@@ -60,11 +69,11 @@ body {
     text-decoration: underline;
-    color: white;
+    color: $color_white;
     margin-right: 5px;
```
can we use the spacing vars here?

jbirori replied:
Yeahh i was planning on coming back and applying the remaining variables in a follow up PR, didnt want to get too in the weeds on this one

MagicWilliams commented on src/styles/styles.scss:
```diff
@@ -82,10 +91,10 @@ body {
     text-decoration: none;
     margin-right: 20px;
```
can we use the spacing vars here?

MagicWilliams commented on src/styles/styles.scss:
```diff
@@ -82,10 +91,10 @@ body {
     text-decoration: none;
     margin-right: 20px;
     font-size: 20px;
```
thoughts on having font-size variables for this font sizing?

jbirori replied:
Ideally we just use the predefined type styles and never need to set font size individually. Down to create variables for this though and then use them within the text styles css definitions 👍🏽

MagicWilliams commented on src/styles/variables.scss:
```diff
@@ -0,0 +1,24 @@
+$spacing_xxx_large: 6rem;
+
+// Colors
+$color_white: #ffffff;
```
idk if we need to prefiz these with `color_` since they're all color names

MagicWilliams commented on src/styles/variables.scss:
```diff
@@ -0,0 +1,24 @@
+// Spacing
+$spacing_xxx_small: 0.5rem;
+$spacing_xx_small: 0.75rem;
+$spacing_x_small: 1rem;
+$spacing_small: 1.25rem;
+$spacing_medium_light: 1.5rem;
+$spacing_medium: 2rem;
+$spacing_medium_plus: 2.5rem;
+$spacing_large: 3rem;
+$spacing_x_large: 4rem;
+$spacing_xx_large: 5rem;
+$spacing_xxx_large: 6rem;
```
I feel like we could/should use less spacing variables? I think using medium light vs medium vs medium plus won't be super noticeable and could undo some of the consistency we're hoping for in the first place. maybe just an xs, sm, md, lg, xl, xxl will give us less options to choose from while giving us enough flexibility to handle all the spacing amounts

jbirori replied:
Yeah for sure, down to cut the list down. Would you mind making a suggestion in the figma file?

MagicWilliams replied:
Gotchuuuu

MagicWilliams reviewed (commented):
a bunch of small stuff I was thinking about - feel free to push back on any of the comments though!

MagicWilliams commented:
I might be missing something too but what's the rationale behind using this Text component vs styling h1 - h6 via the global stylesheet? I hope we move away from adding inline styles like in 

```<Text text="This is the large Body text with red color" style={{color: 'red'}} type="p" />```

and don't see too much difference between that usage and  

```<p className={styles.redText}>This is the large Body text with red color</p>```

jbirori commented:
> I might be missing something too but what's the rationale behind using this Text component vs styling h1 - h6 via the global stylesheet? I hope we move away from adding inline styles like in
> 
> `<Text text="This is the large Body text with red color" style={{color: 'red'}} type="p" />`
> 
> and don't see too much difference between that usage and
> 
> `<p className={styles.redText}>This is the large Body text with red color</p>`

Lol yeah I've been deep in React Native and have lowkey come to love the simplicity of just using a single `Text` component. Down to adjust the PR to just override the h1-h6 tags via css though, probably makes more sense for a web project anyways.

The inline style was just an example to show how one might override styles, i would also prefer to have styles defined via classNames or style objects. Your example would be the same for the Text component as well:

`<Text text="This is the large Body text with red color" style={styles.redText} />`
or if we use a className:
`<Text text="This is the large Body text with red color" className='redText' />`

Will update the CR to just override h1-h6 tags tomorrow

bomanimc commented:
I think one benefit of a text component could be if we have various styles for the same semantic HTML tag. For example, if we have an italics text and a normal text that are both at the p tag level. 

No strong preference on my end though! 

If we do go with the text component approach though, how about passing the context in as a child instead of as a prop? <Text>hello!</Text>. Nice if we need to add span tags and stuff into the text content.

jbirori commented:
> I think one benefit of a text component could be if we have various styles for the same semantic HTML tag. For example, if we have an italics text and a normal text that are both at the p tag level.
> 
> No strong preference on my end though!
> 
> If we do go with the text component approach though, how about passing the context in as a child instead of as a prop? hello!. Nice if we need to add span tags and stuff into the text content.

Ahh yeah, good point on the child suggestion. Definitely more intuitive to pass as children

jbirori committed:
Drop unneccessary prefix from SASS variables, change file names, lowercase class definitions, override the text type element (p, h1 etc..) to match style guide, apply variables to global styles, add font-size variables, refactor Text component to include new prop for mobile type and accept children, and addressed other PR comments..

jbirori commented:
Addressed the PR comments from above! Ultimately decided to just leave the Text component in there as it provides flexibility with the point Bo brought up as we add more text types to our style guide. It's virtually the exact same as using any other text type and accepts all the same props/attributes etc... On the IDE side it also takes advantage of the defined TS types which show up as you're typing:

<img width="237" alt="Screen Shot 2023-01-14 at 7 33 08 PM" src="https://user-images.githubusercontent.com/25678578/212503151-742b4cd4-8c9a-4357-8ebf-62e8b5593375.png">

I noticed on the current home page we change the font size of some of the text for mobile and so to avoid having multiple definitions of the same type in the CSS, I added a media query hook that will change the text type.

Regardless, the default text element styles have been overridden so using `p` or `h1` etc.. will still result in using the correct styles - but for types in our style guide not semantically similar to an HTML tag, like `p+`, they will need to be accompanied with the corresponding className: `<p className="p_large">Text</p>`

jbirori commented:
Inactive for a bit! Gonna merge this in - we can make changes in the future as needed
